File: android/core/src/test/kotlin/me/raddatz/cindy/core/camera/FrameMetricsTests.kt
package me.raddatz.cindy.core.camera

import me.raddatz.cindy.core.signal.BodyPoseObservation
import me.raddatz.cindy.core.signal.PoseJoint
import me.raddatz.cindy.core.signal.PosePoint
import org.junit.Test
import java.nio.ByteBuffer
import kotlin.math.abs
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class FrameMetricsTests {
    /**
     * Y plane whose pixels are `outer`, with the central half set to `inner`. Rows are
     * `rowStride` bytes long; the padding past `width` holds garbage that must never be sampled.
     */
    private fun lumaPlane(width: Int = 64, height: Int = 48, rowStride: Int = width, outer: Int, inner: Int): ByteArray {
        val plane = ByteArray(rowStride * height) { 77 }
        for (y in 0 until height) {
            for (x in 0 until width) {
                val central = y >= height / 4 && y < height * 3 / 4 && x >= width / 4 && x < width * 3 / 4
                plane[y * rowStride + x] = (if (central) inner else outer).toByte()
            }
        }
        return plane
    }

    @Test
    fun lumaSeparatesFrameAndCentre() {
        // Same expectations for a tight plane, a padded one (Android row stride), and a ByteBuffer.
        for (rowStride in listOf(64, 72)) {
            val plane = lumaPlane(rowStride = rowStride, outer = 255, inner = 0)
            val luma = assertNotNull(FrameMetricsCalculator.luma(plane, 64, 48, rowStride))
            assertEquals(0f, luma.center)
            // The central half covers a quarter of the sampled pixels.
            assertTrue(abs(luma.mean - 0.75f) < 0.01f)

            val direct = ByteBuffer.allocateDirect(plane.size).put(plane).also { it.flip() }
            val buffered = FrameMetricsCalculator.measure(direct, 64, 48, rowStride)
            assertEquals(luma.mean, buffered.lumaMean)
            assertEquals(luma.center, buffered.lumaCenter)
            assertEquals(0, direct.position())
        }
        assertNull(FrameMetricsCalculator.luma(ByteArray(0), 0, 0, 0))
    }

    @Test
    fun shoulderWidthIsTheDistanceBetweenShoulders() {
        val pose = BodyPoseObservation(
            mapOf(
                PoseJoint.LEFT_SHOULDER to PosePoint(0.2f, 0.5f, 0.9f),
                PoseJoint.RIGHT_SHOULDER to PosePoint(0.5f, 0.9f, 0.9f),
            ),
        )
        assertTrue(abs((pose.shoulderWidth ?: 0f) - 0.5f) < 1e-6f)
        assertNull(BodyPoseObservation(mapOf(PoseJoint.NOSE to PosePoint(0.5f, 0.5f, 1f))).shoulderWidth)
    }
}

File: CindyTests/FrameMetricsTests.swift
import CoreVideo
import Foundation
import Testing
@testable import Cindy

struct FrameMetricsTests {
    /// Full-range 4:2:0 buffer whose Y plane is `outer`, with the central half set to `inner`.
    private func lumaBuffer(width: Int = 64, height: Int = 48, outer: UInt8, inner: UInt8) throws -> CVPixelBuffer {
        var buffer: CVPixelBuffer?
        CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_420YpCbCr8BiPlanarFullRange,
                            nil, &buffer)
        let pixelBuffer = try #require(buffer)
        CVPixelBufferLockBaseAddress(pixelBuffer, [])
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }
        let base = try #require(CVPixelBufferGetBaseAddressOfPlane(pixelBuffer, 0)).assumingMemoryBound(to: UInt8.self)
        let bytesPerRow = CVPixelBufferGetBytesPerRowOfPlane(pixelBuffer, 0)
        for y in 0..<height {
            for x in 0..<width {
                let central = y >= height / 4 && y < height * 3 / 4 && x >= width / 4 && x < width * 3 / 4
                base[y * bytesPerRow + x] = central ? inner : outer
            }
        }
        return pixelBuffer
    }

    @Test func lumaSeparatesFrameAndCentre() throws {
        let buffer = try lumaBuffer(outer: 255, inner: 0)
        let luma = try #require(FrameMetricsCalculator.luma(of: buffer))
        #expect(luma.center == 0)
        // The central half covers a quarter of the sampled pixels.
        #expect(abs(luma.mean - 0.75) < 0.01)
    }

    @Test func shoulderWidthIsTheDistanceBetweenShoulders() {
        let pose = BodyPoseObservation(joints: [
            .leftShoulder: PosePoint(x: 0.2, y: 0.5, confidence: 0.9),
            .rightShoulder: PosePoint(x: 0.5, y: 0.9, confidence: 0.9),
        ])
        #expect(abs((pose.shoulderWidth ?? 0) - 0.5) < 1e-6)
        #expect(BodyPoseObservation(joints: [.nose: PosePoint(x: 0.5, y: 0.5, confidence: 1)]).shoulderWidth == nil)
    }
}
